### PR TITLE
#342 Implemented restart policy

### LINKED ISSUE #342: Implement a restart endpoint
**Is your feature request related to a problem? Please describe.**
To allow users to restart their workloads, cyclops-ctrl should expose a restart endpoint. You can register the endpoint and its handler function [here](https://github.com/cyclops-ui/cyclops/blob/5668c95e19d4e39a5c6130ca4245a497095fe861/cyclops-ctrl/internal/handler/handler.go#L83).

The endpoint should receive the group, version, kind, name and the namespace of the resource that needs to be restarted. You can fetch those similarly to how it was done [here](https://github.com/cyclops-ui/cyclops/blob/5668c95e19d4e39a5c6130ca4245a497095fe861/cyclops-ctrl/internal/controller/modules.go#L562). You can also check this link on how to restart a resource using go-client [here](https://stackoverflow.com/questions/61335318/how-to-restart-a-deployment-in-kubernetes-using-go-client).

Scope of the issue is implementing **only** the backend endpoint. Frontend part will be implemented separately.

### PR DESCRIPTION
**Pull request by patelajay745**
closes #342 

## 📑 Description

Implemented restart endpoint. The endpoint url is resources/restart?group=apps&version=v1&kind=Deployment&name=<name-deploy>&namespace=<namespace> 

## ✅ Checks
- [ ] I have updated the documentation as required
- [X] I have performed a self-review of my code

## ℹ Additional context

Some SS for reference.

Before requesting endpoint
Screenshot 2024-07-10 at 8 13 03 AM

Endpoint calling 
Screenshot 2024-07-10 at 8 13 29 AM

 After requesting endpoint
 
Screenshot 2024-07-10 at 8 13 47 AM

If there is any error
Screenshot 2024-07-10 at 8 22 15 AM

### CODE REVIEW

**Comment by patelajay745** [comment]
Can you please check code and update me?

File: cyclops-ctrl/internal/cluster/k8sclient/client.go [hunk 1]
```diff
@@ -246,6 +248,21 @@ func (k *KubernetesClient) GetStatefulSetsLogs(namespace, container, name string
...
+		return err
+	}
+
+	if deploy.Spec.Template.ObjectMeta.Annotations == nil {
+		deploy.Spec.Template.ObjectMeta.Annotations = make(map[string]string)
+	}
+	deploy.Spec.Template.ObjectMeta.Annotations["kubectl.kubernetes.io/restartedAt"] = time.Now().Format(time.RFC3339)
+
+	_, err = k.clientset.AppsV1().Deployments(namespace).Update(context.Background(), deploy, metav1.UpdateOptions{})
+	return err
```
Comment (petar-cvit) [inline cyclops-ctrl/internal/cluster/k8sclient/client.go:263]: Can you also implement this for statefulsets and daemonsets?

File: cyclops-ctrl/internal/controller/modules.go [hunk 2]
```diff
@@ -577,6 +577,27 @@ func (m *Modules) GetManifest(ctx *gin.Context) {
 	ctx.String(http.StatusOK, manifest)
 }
 
+func (m *Modules) GetRestarted(ctx *gin.Context) {
```
Comment (petar-cvit) [inline cyclops-ctrl/internal/controller/modules.go:580]: Can we rename this function to `Restart` and make the method `POST`? Also can we then read group, version, kind, name and namespace from the request body?

**Review by petar-cvit — CHANGES_REQUESTED** [review]

File: cyclops-ctrl/internal/controller/modules.go (hunk 2 above)
Comment (patelajay745) [reply cyclops-ctrl/internal/controller/modules.go]: Sure. Changed.

File: cyclops-ctrl/internal/cluster/k8sclient/client.go (hunk 1 above)
Comment (patelajay745) [reply cyclops-ctrl/internal/cluster/k8sclient/client.go:262]: Yes. I have implemented Statefulsets and Daemonsets in new commits. Please review and update me.

File: cyclops-ctrl/internal/cluster/k8sclient/client.go [hunk 3]
```diff
@@ -273,6 +317,19 @@ func (k *KubernetesClient) GetManifest(group, version, kind, name, namespace str
 	return string(data), nil
 }
 
+func (k *KubernetesClient) Restart(group, version, kind, name, namespace string) (any, error) {
```
Comment (petar-cvit) [inline cyclops-ctrl/internal/cluster/k8sclient/client.go:320]: You can remove any from the return 
```suggestion
func (k *KubernetesClient) Restart(group, version, kind, name, namespace string) error {
```

File: cyclops-ctrl/internal/cluster/k8sclient/client.go [hunk 4]
```diff
@@ -273,6 +317,19 @@ func (k *KubernetesClient) GetManifest(group, version, kind, name, namespace str
...
+	switch {
+	case isDeployment(group, version, kind):
+		return nil, k.RestartDeployment(group, version, kind, name, namespace)
+	case isDaemonSet(group, version, kind):
+		return nil, k.RestartDaemonsets(group, version, kind, name, namespace)
+	case isStatefulSet(group, version, kind):
+		return nil, k.RestartStatefulSets(group, version, kind, name, namespace)
+	}
+
+	return nil, nil
```
Comment (petar-cvit) [inline cyclops-ctrl/internal/cluster/k8sclient/client.go:330]: Can we throw an error if provided group, version and kind cannot be restarted

```suggestion
	return errors.New(fmt.Sprintf("cannot restart: %v/%v %v %v/%v", group, version, kind, namespace, name))
```

**Review by petar-cvit — CHANGES_REQUESTED** [review]
Nice job @patelajay745 , work for all three now. Left two more comments

File: cyclops-ctrl/internal/cluster/k8sclient/client.go (hunk 3 above)
Comment (patelajay745) [reply cyclops-ctrl/internal/cluster/k8sclient/client.go]: removed

File: cyclops-ctrl/internal/cluster/k8sclient/client.go (hunk 4 above)
Comment (patelajay745) [reply cyclops-ctrl/internal/cluster/k8sclient/client.go]: added

**Review by petar-cvit — APPROVED** [review]
Thanks @patelajay745 🧡

**Pull request merged**